File: src/main/java/com/everyonewaiter/application/image/ImageService.java
package com.everyonewaiter.application.image;

import static java.util.Objects.requireNonNull;
import static java.util.Objects.requireNonNullElse;
import static org.slf4j.LoggerFactory.getLogger;

import com.everyonewaiter.application.image.provided.ImageManager;
import com.everyonewaiter.application.image.required.ImageClient;
import com.everyonewaiter.application.image.required.ImageFormatConverter;
import com.everyonewaiter.application.image.required.PDFToImageConverter;
import com.everyonewaiter.domain.image.FailedUploadImageException;
import com.everyonewaiter.domain.image.UnsupportedFileTypeException;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;

@Component
@RequiredArgsConstructor
class ImageService implements ImageManager {

  private static final Logger LOGGER = getLogger(ImageService.class);

  private final ImageClient imageClient;
  private final ImageFormatConverter imageFormatConverter;
  private final PDFToImageConverter pdfToImageConverter;

  @Override
  public String upload(String prefix, MultipartFile file) {
    MultipartFile imageFile = convertFileToImage(file, prefix);
    String imageFileName = requireNonNull(imageFile.getOriginalFilename());
    File tempFile = new File(imageFileName.replace("/", "."));

    try {
      imageFile.transferTo(tempFile);

      imageClient.upload(tempFile, imageFileName, imageFile.getContentType());

      return imageFileName;
    } catch (IOException exception) {
      LOGGER.error("이미지 업로드 중 임시 파일 생성에 실패하였습니다. {}", imageFileName, exception);

      throw new FailedUploadImageException();
    } finally {
      deleteTempFile(tempFile);
    }
  }

  private MultipartFile convertFileToImage(MultipartFile file, String prefix) {
    String contentType = requireNonNullElse(file.getContentType(), "empty");

    return switch (contentType) {
      case MediaType.IMAGE_PNG_VALUE,
           MediaType.IMAGE_JPEG_VALUE -> imageFormatConverter.convertToWebp(file, prefix);
      case MediaType.APPLICATION_PDF_VALUE -> pdfToImageConverter.convertFirstPage(file, prefix);
      default -> throw new UnsupportedFileTypeException();
    };
  }

  @Override
  public void delete(String imageName) {
    imageClient.delete(imageName);
  }

  private void deleteTempFile(File tempFile) {
    try {
      Files.deleteIfExists(tempFile.toPath());
    } catch (IOException exception) {
      LOGGER.error("파일 삭제에 실패하였습니다. {}", tempFile.getName());
    }
  }

}

File: src/main/kotlin/com/everyonewaiter/application/image/service/ImageService.kt
package com.everyonewaiter.application.image.service

import com.everyonewaiter.application.image.dto.ImageUpload
import com.everyonewaiter.common.file.converter.ImageFormatConverter
import com.everyonewaiter.common.file.converter.PdfFileConverter
import com.everyonewaiter.common.file.extension.isImage
import com.everyonewaiter.common.file.extension.isPdf
import com.everyonewaiter.global.exception.BusinessException
import com.everyonewaiter.global.exception.ErrorCode
import com.everyonewaiter.infrastructure.image.ImageClient
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile

@Service
class ImageService(
    private val imageClient: ImageClient,
    private val imageFormatConverter: ImageFormatConverter,
    private val pdfFileConverter: PdfFileConverter,
) {
    fun upload(request: ImageUpload.Request): String {
        val file = convertToWebp(request.file, request.prefix)
        return imageClient.upload(file)
    }

    private fun convertToWebp(
        file: MultipartFile,
        prefix: String?,
    ): MultipartFile =
        when {
            file.isPdf -> {
                val image = pdfFileConverter.firstPageToImage(file, prefix).getOrNull()
                image ?: throw BusinessException(ErrorCode.FAILED_CONVERT_PDF_TO_IMAGE)
            }

            file.isImage -> {
                val image = imageFormatConverter.convertToWebp(file, prefix).getOrNull()
                image ?: throw BusinessException(ErrorCode.FAILED_CONVERT_IMAGE_FORMAT)
            }

            else -> throw BusinessException(ErrorCode.ALLOW_IMAGE_AND_PDF_FILE)
        }
}
